ExprWithLimits `(free|bound)_symbols` updated

smichr opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed

When a user-defined function is acting as a bound variable, it's arguments are implicitly bound and changes have been made to reflect this by modifying `bound_symbols` and `free_symbols` to account for those via change to `ExprWithLimits`.

`Basic.canonical_variables` will now include symbols contained within functions acting as bound symbols.

These changes are reflected in expressions returned for `as_dummy`:

```python
>>> f=Function('f')
>>> Integral(x,(f(x), y)).as_dummy()
Integral(_1, (_0, y))  <----- formerly Integral(x, (_0, y))
```

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* core
  * `canonical_variables` now includes free symbols from the arguments of bound functions
* concrete
  * symbols in functions appearing as bound functions are now considered bound, too
  * ExprWithLimits (like Integral) now reports arguments of functions as free symbols when appearing as singleton limit
  * Integrals like `Integral(x, f(x))` will not treat `x` and independent of `f(x)` and will not respond to `.doit()`
<!-- END RELEASE NOTES -->

smichr committed:
bound functions bind their free symbols

smichr committed:
symbols in bound functions not free

smichr committed:
canonical variables includes arg variables

smichr committed:
20809: tests for as_dummy

smichr commented:
ping @oscarbenjamin @jksuom

oscarbenjamin commented:
This doesn't seem consistent to me:
```python
In [1]: integral = Integral(x,(f(x), y))

In [2]: integral
Out[2]: 
y          
⌠          
⎮ x d(f(x))
⌡          
           

In [3]: integral.doit()
Out[3]: x⋅y

In [4]: integral.as_dummy()
Out[4]: 
y         
⌠         
⎮ _1 d(_0)
⌡         
          

In [5]: integral.free_symbols
Out[5]: {y}

In [6]: integral.doit().free_symbols
Out[6]: {x, y}
```
I think that `free_symbols` should be the same for `integral` and `integral.doit()`.

smichr commented:
> I think that `free_symbols` should be the same for `integral` and `integral.doit()`.

As you point out on the issue page, this is a nonsensical integral. Like `Poly` raising an error for `Poly(x + sqrt(x), x)` I think we should not evaluate an integral written in terms of `f(x)` when `x` appears independently.

oscarbenjamin commented:
The situation where this kind of integral comes up is something like separation of variables in `dsolve`. Mathematically we write something like this:
```
dx/dt = f(x)*g(t)
int 1/f(x) dx = int g(t) dt
```
The problem is that in sympy the derivative dx/dt only makes sense if x is a function `x(t)` so then it bcomes:
```
t = Symbol('t')
x = Function('x')
Derivative(x(t), t) = f(x(t))*g(t)
Integral(1/f(x(t)), x(t)) = Integral(g(t), t)
```
Here in these integrals we definitely want to keep `x(t)` and `t` separate but I'm not sure how generally that applies.

I think we also want to preserve the idea that an indefinite integral is an antiderivative so those need to be consistent. Situations like `Derivative(f(x(t), t), x(t))` come up a lot in dynamics problems e.g. in the Lagrange equations. Those should be understood as partial derivatives where `t` is a constant when differentiating wrt `x(t)`. Maybe these problem stem from needing to express a derivative wrt a variable rather than a numbered argument of a function.

smichr committed:
Integral(x, y).bound_variables -> [y]

The result of as_dummy for the evaluate-at
integral written as a limit of length 1 now
is written in the lenght-2 format to show
that something is both bound and free.

This singleton limit expression will now
be listed as a bound variable, too.

`doit` will not return an evaluated integral
when a function has symbols and functions with
any of those symbols in the bound symbols. e.g.
Integral(x + y, y) is not evaluated if y is f(x)
just like it is not evaluated when x = f(y).

smichr commented:
@oscarbenjamin , please see commit message on last commit about when an integral will not evaluate.

smichr commented:
Hmm...guess that piece of code was uncovered before. Not sure if it is the explication of the eval-at that caused the code path to be traversed or not. I will try to add a direct test of this path.

smichr committed:
watch for (a and b) is None

This actually needs to be tested as None in (a, b) or similar
since (0 and None) == 0, not None.

smichr commented:
ping @oscarbenjamin -- this is passing now and the objective for the doit and non-doit forms is attained by not doing an integral that does not contain enough information. From the commit message:

> Integral(x + y, y) is not evaluated if y is f(x) just like it is not evaluated when x = f(y).

An error in handling None was also fixed and I searched for `\(.* and .*\) is None` and did not find any other instances.

oscarbenjamin commented on sympy/core/basic.py:
```diff
@@ -582,9 +582,8 @@ def canonical_variables(self):
-                while d.name in names:
-                    d = next(dums)
+            while d.name in names:
+                d = next(dums)
```
What happens if StopIteration is raise here?

smichr replied:
It's an infinite generator so I can't imagine a scenario for this.

oscarbenjamin commented:
This still seems inconsistent to me:
```python
In [5]: integral.free_symbols
Out[5]: {y}

In [6]: integral.doit().free_symbols
Out[6]: {x, y}
```
Is the intention of this PR that we should also change what `Integral.doit` does so the integral remains unevaluated or raises an error? If so I think I would want to see the effect of that change before changing free symbols to match it. Are there many places in the codebase that depend on this?

smichr committed:
leave unknown undone

smichr commented:
> This still seems inconsistent to me:

hmmm...must have gotten lost in the shuffle. I pushed a change and will have to see the impact now on the rest of the suite.

smichr commented on sympy/integrals/integrals.py:
```diff
@@ -397,6 +401,11 @@ def doit(self, **hints):
             return self
+        bv = [f[0] for f in self.limits if not f[0].is_Symbol]
+        if bv:
+            free = self.function.free_symbols - set(bv)
```
```suggestion
            free = self.function.expr_free_symbols - set(bv)
```

smichr commented on sympy/integrals/integrals.py:
```diff
@@ -397,6 +401,11 @@ def doit(self, **hints):
             return self
+        bv = [f[0] for f in self.limits if not f[0].is_Symbol]
+        if bv:
+            free = self.function.free_symbols - set(bv)
+            if any(i.free_symbols & free for i in bv):
```
```suggestion
            free = self.function.expr_free_symbols - set(bv)
            if free and any(i.expr_free_symbols & free for i in bv):
```

smichr committed:
watch for expr_free_symbols vs free_symbols

smichr commented on sympy/integrals/integrals.py:
```diff
@@ -397,6 +401,11 @@ def doit(self, **hints):
         from sympy.concrete.summations import Sum
         if not hints.get('integrals', True):
             return self
+        bv = [f[0] for f in self.limits if not f[0].is_Symbol]
```
```suggestion
        bv = [f[0] for f in self.limits if f[0].expr_free_symbols != {f[0]}]
```

smichr commented on sympy/integrals/integrals.py:
```diff
@@ -397,6 +401,11 @@ def doit(self, **hints):
         from sympy.concrete.summations import Sum
         if not hints.get('integrals', True):
             return self
+        bv = [f[0] for f in self.limits if f[0].expr_free_symbols != {f[0]}]
```
```suggestion
        bv = [f[0] for f in self.limits if (
            f[0].expr_free_symbols != {f[0]} and
            f[0].free_symbols != {f[0]}):
```

smichr commented on sympy/integrals/integrals.py:
```diff
@@ -397,6 +401,13 @@ def doit(self, **hints):
             return self
+        bv = [f[0] for f in self.limits if (
+            f[0].expr_free_symbols != {f[0]} and
+            f[0].free_symbols != {f[0]}):
+        if bv:
+            free = self.function.expr_free_symbols - set(bv)
```
```suggestion
            f[0].free_symbols != {f[0]})]:
        if bv:
            bv = set(bv)
            free = self.function.expr_free_symbols - bv
```

smichr commented on sympy/integrals/integrals.py:
```diff
@@ -397,6 +401,14 @@ def doit(self, **hints):
             return self
+        bv = [f[0] for f in self.limits if (
+            f[0].expr_free_symbols != {f[0]} and
+            f[0].free_symbols != {f[0]})]:
```
```suggestion
            f[0].free_symbols != {f[0]})]
```

smichr commented:
> I think I would want to see the effect of that change before changing free symbols to match it. Are there many places in the codebase that depend on this?

The `free_symbols` for `Integral` has only been changed to consider free symbols of the variables of integration. So instead of `f(x)` alone being added when it is the dummy variable, now `x` is added, too. But the change to not evaluate an integral with `f(x)` as a dummy variable if there is a free `x` in the integrand is now enforced and appears to have no impact on the existing suite.

oscarbenjamin commented:
I wonder if we're just making everything complicated by allowing functions to be integration variables at all.

I don't know whether we should really reserve that kind of usage for something like a Riemann-Stieltjes integral:
https://en.wikipedia.org/wiki/Riemann%E2%80%93Stieltjes_integral

smichr commented:
> everything complicated by allowing functions to be integration variables at all

The complication will still be there when we allow Indexed and Normal as integration variables. There may be others, but I at least tripped over those two which behave differently than AppliedUndef.

oscarbenjamin commented:
> The complication will still be there when we allow Indexed and Normal as integration variables.

Do we need to allow these? I think we do need to be able to define somehow what is allowed and exactly what things should be considered "symbol-like" and make sure that it is possible to write code that can handle symbol-like objects without needing to special case these types.

smichr commented:
> Do we need to allow these?

Maybe we should just make it a printing issue so Subs(Integral(x, x), (x, f(x))) prints as `Integral(f(x), f(x))` and entry of the later generates the former automatically. So the user would never see this. But maybe this is a bad idea since such things are public. 

Maybe a private attribute could hold what the integration variable is. The private variable could be used for representation but when working internally with the integral it would be easier to not have to worry about running into non-symbols. Things like `bound_symbols` and `free_symbols` would still have to take the value into consideration, however, since something like `integrate(x + _, _); _ = f(x)` should still return an unevaluated `Integral`.

Nagging suspicion that this really doesn't solve anything since you still have to look at what the integration variables is when making decisions about `subs`, for example. 

So what can be an integration variable? We hack for this by defining `_diff_wrt`:
```python
>>> from sympy.stats import *
>>> X = Normal('X', 2, 4)
>>> X._diff_wrt
True
```

Maybe we need to make that a public attribute and decide on a good name -- maybe `.is_symbol` in spirit of `is_number` or `is_integer` being different than `is_Number` and `is_Integer`.

smichr commented:
> code that can handle symbol-like objects without needing to special case these types.

note that the only special attention given in this PR is to check for when something is not `is_Symbol` -- so `is_Symbol` being true indicates the base case.

smichr commented:
Getting a better definition of what is and is not allowed as a variable of integration can go in another PR. This one at least resolves issues with getting `as_dummy` to correctly identify the dummy variables and stops itnegrals with symbol-dependent variables from evaluating if that symbol is also free in the integrand.

Will commit in a few hours unless there is objection.

oscarbenjamin commented:
I'm not sure about the way `at_dummy` loses the information relating the different "symbols":
```python

In [10]: I = Integral(x + f(x), (f(x), 0, f(x)))

In [11]: I
Out[11]: 
f(x)                   
 ⌠                     
 ⎮   (x + f(x)) d(f(x))
 ⌡                     
 0                     

In [12]: I.doit()
Out[12]: 
f(x)                   
 ⌠                     
 ⎮   (x + f(x)) d(f(x))
 ⌡                     
 0                     

In [13]: I.as_dummy().doit()
Out[13]: 
           2   
          f (x)
_1⋅f(x) + ─────
            2 
```
Ideally we would have an invariant like `I.doit() == I.as_dummy().doit()`.

On master we have:
```python
In [1]: I = Integral(x + f(x), (f(x), 0, f(x)))

In [2]: I.as_dummy()
Out[2]: 
_0               
⌠                
⎮  (_0 + x) d(_0)
⌡                
0                

In [3]: I.as_dummy().doit()
Out[3]: 
  2       
_0        
─── + _0⋅x
 2        

In [4]: I.doit()
Out[4]: 
          2   
         f (x)
x⋅f(x) + ─────
           2  
```
That also seems wrong because the dummy should be gone after the `doit`.

smichr commented:
> That also seems wrong because the dummy should be gone after the `doit`.

You are asking `as_dummy` to do more than it was intended to do. The intention was to give a canonical expression -- not necessarily the same mathematically -- of the original expression for purposes of comparison. It should always be true that two expressions that give the same result except for names of variables will be the same when expressed in `as_dummy` form. It's possible to make `as_dummy` be the same as the original but with different names; this requires recreating the class of object that is being replaced with a Symbol. 

So is this a to be considered a bug or a feature?

oscarbenjamin commented:
I guess what you're saying is that the invariant should be that if the doits can evaluate then
```
e1.doit() == e2.doit()   <-->   e1.as_dummy() == e2.as_dummy()
```

smichr commented:
> that if the doits can evaluate then

yes; `as_dummy` just yields a canonical expression that can be tested modulo free symbols. Using `doit` on the `as_dummy` result would be like using `evalf` twice on an expression in that the precision of the second instance would no longer reflect the precision -- actual precision -- of the first result.

oscarbenjamin commented:
Okay how about this:
```python
In [18]: I1 = Integral(x, (x, y), (y, x), (x, y))

In [19]: I2 = Integral(x, (x, y), (y, z), (z, y))

In [20]: I1.doit() == I2.doit()
Out[20]: True

In [21]: I1.as_dummy() == I2.as_dummy()
Out[21]: False

In [22]: I1.as_dummy()
Out[22]: 
y _0 _1                     
⌠ ⌠  ⌠                      
⎮ ⎮  ⎮  _0 d(_0) d(_1) d(_0)
⌡ ⌡  ⌡                      
                            

In [23]: I2.as_dummy()
Out[23]: 
y _2 _1                     
⌠ ⌠  ⌠                      
⎮ ⎮  ⎮  _0 d(_0) d(_1) d(_2)
⌡ ⌡  ⌡ 
```

smichr commented:
> Okay how about this:

I agree. They should be the same. I'll take a look.

smichr committed:
minimize canonical variable count

smichr commented:
This was stress tested with
```python
v = var('a:d')
lim = list(subsets(v,repetition=True))
for i in subsets(v, repetition=True):
 if not i:continue
 i = Add(*i)
 for j in subsets(lim, repetition=True):
  j = [k for k in j if k]
  if not j:continue
  I = Integral(i,*j)
  try:assert I.doit() == I.as_dummy().doit()
  except:
   a,b,c=(I,I.doit(),I.as_dummy())
   assert None, a
```

smichr commented:
tricky wrangling needed; closing until done

smichr committed:
override canonical_variables for expr_with_limits

smichr committed:
wip

smichr commented on sympy/core/tests/test_basic.py:
```diff
@@ -299,14 +299,63 @@ def test_as_dummy():
+        ) == Integral(_1, _0, (_0, y), (_1, x))
+    assert Integral(x, y, z, (x, x)).as_dummy(
+        ) == Integral(_2, (_0, y), (_1, z), (_2, x))
+    v = symbols('a:d')
+
+
+def test_stress():
+    from sympy.utilities.iterables import subsets
+    lim = list(subsets(v,repetition=True))
+    for i in subsets(v, repetition=True):
+     if not i:continue
+     i = Add(*i)
+     for j in subsets(lim, repetition=True):
+      j = [k for k in j if k]
+      if not j:continue
+      I = Integral(i,*j)
+      try:assert I.doit() == I.as_dummy().doit()
+      except:
+       a,b,c=(I,I.doit(),I.as_dummy())
+       assert None, a
```
```suggestion


def test_stress():
    from sympy.utilities.iterables import subsets
    v = symbols('a:d')
    lim = list(subsets(v,repetition=True))
    for i in subsets(v, repetition=True):
     if not i:continue
     i = Add(*i)
     for j in subsets(lim, repetition=True):
      j = [k for k in j if k]
      if not j:continue
      I = Integral(i,*j)
      if I.doit() != I.as_dummy().doit():
       assert None, (I, I.doit())
```

smichr committed:
Update sympy/core/tests/test_basic.py

smichr commented on sympy/concrete/expr_with_limits.py:
```diff
@@ -282,17 +291,89 @@ def free_symbols(self):
+        ========
+
+        >>> from sympy import Integral, Function
+        >>> from sympy.abc import x, y, z
```
```suggestion
        >>> from sympy.abc import x, y
```

smichr commented on sympy/core/tests/test_basic.py:
```diff
@@ -299,14 +299,61 @@ def test_as_dummy():
+
+
+def test_stress():
+    from sympy.utilities.iterables import subsets
```
```suggestion
    from sympy import Add
    from sympy.utilities.iterables import subsets
```

smichr committed:
Apply suggestions from code review

smichr commented on sympy/concrete/expr_with_limits.py:
```diff
@@ -282,17 +291,89 @@ def free_symbols(self):
+
+        A new dummy was not needed for variable z so the ``_0``
+        was re-used.
+
+        >>> Integral(x, (x, y), (y, z), z).as_dummy()
+        Integral(_0, (_0, _1), (_1, _0), (_0, z))
+        """
+        reps = super().canonical_variables
+        if len(self.limits) <= 2:
+            return reps
+        _0 = reps[self.limits[0][0]]
+        free = self.function.free_symbols
+        for i, ui in enumerate(uniq(i[0] for i in self.limits)):
+            if i > 1 and ui not in free:
+                reps[ui] = _0
+        return reps
```
```suggestion
        """
        reps = super().canonical_variables
        if len(reps) <= 2:
            return reps
        _0 = reps[self.limits[0][0]]
        free = self.function.free_symbols
        for i, ui in enumerate(uniq(i[0] for i in self.limits)):
            if i > 1 and ui not in free:
                reps[ui] = _0
        return reps
```

smichr committed:
Update sympy/concrete/expr_with_limits.py

smichr commented on sympy/concrete/expr_with_limits.py:
```diff
@@ -282,17 +291,89 @@ def free_symbols(self):
+        _0 = reps[self.limits[0][0]]
+        free = self.function.free_symbols
+        for i, ui in enumerate(uniq(i[0] for i in self.limits)):
+            if i > 1 and ui not in free:
```
```suggestion
            if i > 1 and ui in reps:
```

smichr commented on sympy/core/tests/test_basic.py:
```diff
@@ -299,14 +299,62 @@ def test_as_dummy():
+      if not j:continue
+      I = Integral(i,*j)
+      if I.doit() != I.as_dummy().doit():
+       assert None, (I, I.doit())
```
```suggestion
       assert None, (I, I.as_dummy())
```

smichr committed:
Apply suggestions from code review

smichr commented on sympy/concrete/expr_with_limits.py:
```diff
@@ -282,17 +291,86 @@ def free_symbols(self):
+        if len(reps) <= 2:
+            return reps
+        _0 = reps[self.limits[0][0]]
+        free = self.function.free_symbols
```
```suggestion
```

smichr committed:
Apply suggestions from code review
